### PR TITLE
Adding TLS to server REST API

### PR DESCRIPTION
**Pull request by gennadylaventman**
Adding TLS and fixing small glitch in comm.httptransport.go

Signed-off-by: Gennady Laventman <gennady@il.ibm.com>

### CODE REVIEW

File: pkg/server/server_test.go [hunk 1]
```diff
@@ -41,6 +42,7 @@ type serverTestEnv struct {
 	adminSigner    crypto.Signer
 	client         *mock.Client
 	certCol        *certificateauthority.CACertCollection
+	tlsCertCol     *certificateauthority.CACertCollection
```
Comment (tock-ibm) [inline pkg/server/server_test.go:45]: rename them such that it is clear which is for client-comm and which is for cluster-comm

File: pkg/server/server_test.go [hunk 2]
```diff
@@ -623,8 +680,122 @@ func TestServerWithFailureScenarios(t *testing.T) {
...
+	require.NoError(t, err)
+
+	err = verifier.Verify(userBytes, user.GetSignature())
+	require.NoError(t, err)
+
+	require.True(t, user.GetResponse().GetUser() != nil &&
+		user.GetResponse().GetUser().GetId() == "testUser")
+}
+
+func TestSyncUserTxWithServerAndClientTLS(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (gennadylaventman) [reply pkg/server/server_test.go:740]: Yep

File: deployment/config-docker/config.yml [hunk 3]
```diff
@@ -41,6 +41,12 @@ server:
     block: 100
   # logLevel can be debug, info, warn, err, and panic
   logLevel: info
+  tls:
+    enabled: false
+    # X.509 certificate used for TLS server
+    serverCertificatePath: ./testdata/tls/server/server.cert
+    # Private key for TLS server
+    serverKeyPath: ./testdata/tls/server/server.key
```
[the remark this answers is not in the document]
Comment (gennadylaventman) [reply deployment/config-docker/config.yml:49]: Done and done

File: pkg/server/server_test.go (hunk 1 above)
Comment (gennadylaventman) [reply pkg/server/server_test.go:45]: One is for signing, second is for sdk-server tls comm, none is for cluster comm

File: pkg/server/testutils/crypto_utils.go [hunk 4]
```diff
@@ -215,21 +215,13 @@ func GenerateTestClientCrypto(t *testing.T, names []string, withIntermediateCA .
 	}
 
 	for _, name := range names {
-		pemCert, privKey, err := IssueCertificate("BCDB Client "+name, "127.0.0.1", keyPair)
```
[the remark this answers is not in the document]
Comment (gennadylaventman) [reply pkg/server/testutils/crypto_utils.go:218]: Sounds ok. Will do.

[the remark this answers is not in the document]
Comment (gennadylaventman) [reply pkg/server/testutils/crypto_utils.go:218]: I will change it to `GenerateTestCrypto` and his counterpart to `LoadTestCrypto`

File: pkg/certificateauthority/ca_chain_test.go [hunk 5]
```diff
@@ -3,19 +3,20 @@
 package certificateauthority
 
 import (
-	"github.com/hyperledger-labs/orion-server/config"
 	"path"
 	"testing"
 
+	"github.com/hyperledger-labs/orion-server/config"
+
```
Comment (tock-ibm) [inline pkg/certificateauthority/ca_chain_test.go:10]: goimports

File: pkg/server/server_test.go [hunk 6]
```diff
@@ -117,40 +119,22 @@ func newServerTestEnv(t *testing.T) *serverTestEnv {
...
-	_, err = serverRootCACertFile.Write(rootCAPemCert)
-	require.NoError(t, err)
-	require.NoError(t, serverRootCACertFile.Close())
-
-	pemCert, privKey, err := testutils.IssueCertificate("BCDB Instance", "127.0.0.1", keyPair)
+	err = os.WriteFile(path.Join(tempDir, "serverRootCACert.pem"), rootCAPemCert, 0666)
 	require.NoError(t, err)
 
-	pemCertFile, err := os.Create(path.Join(tempDir, "server.pem"))
+	pemCert, privKey, err := testutils.IssueCertificate("BCDB Instance", "127.0.0.1", caKeyPair)
```
Comment (tock-ibm) [inline pkg/server/server_test.go:125]: Orion Instance

File: pkg/server/server_test.go [hunk 7]
```diff
@@ -160,6 +144,43 @@ func newServerTestEnv(t *testing.T) *serverTestEnv {
 	)
 	require.NoError(t, err)
 
+	if serverTLS {
+		tlsRootCAPemCert, tlsCaPrivKey, err := testutils.GenerateRootCA("BCDB TLS RootCA", "127.0.0.1")
```
Comment (tock-ibm) [inline pkg/server/server_test.go:148]: BCDB -> Orion

File: pkg/server/server_test.go [hunk 8]
```diff
@@ -160,6 +144,43 @@ func newServerTestEnv(t *testing.T) *serverTestEnv {
...
+		tlsCertsCollection, err := certificateauthority.NewCACertCollection([][]byte{block.Bytes}, nil)
+		require.NoError(t, err)
+
+		err = tlsCertsCollection.VerifyCollection()
+		require.NoError(t, err)
+
+		err = os.WriteFile(path.Join(tempDir, "tlsServerRootCACert.pem"), tlsRootCAPemCert, 0666)
+		require.NoError(t, err)
+
+		tlsPemCert, tlsPrivKey, err := testutils.IssueCertificate("BCDB TLS Instance", "127.0.0.1", tlsCAKeyPair)
```
Comment (tock-ibm) [inline pkg/server/server_test.go:167]: Orion Server TLS Instance

File: pkg/server/server_test.go [hunk 9]
```diff
@@ -160,6 +144,43 @@ func newServerTestEnv(t *testing.T) *serverTestEnv {
...
+		require.NoError(t, err)
+		err = os.WriteFile(path.Join(tempDir, "tlsServer.pem"), tlsPemCert, 0666)
+		require.NoError(t, err)
+		err = os.WriteFile(path.Join(tempDir, "tlsServer.key"), tlsPrivKey, 0666)
+		require.NoError(t, err)
+		if clientTLS {
+			tlsPemCert, tlsPrivKey, err := testutils.IssueCertificate("BCDB Client TLS Instance", "127.0.0.1", tlsCAKeyPair)
```
Comment (tock-ibm) [inline pkg/server/server_test.go:174]: Orion

File: pkg/server/server_test.go [hunk 10]
```diff
@@ -623,8 +680,122 @@ func TestServerWithFailureScenarios(t *testing.T) {
 	}
 }
 
+func TestSyncUserTxWithServerTLS(t *testing.T) {
```
Comment (tock-ibm) [inline pkg/server/server_test.go:683]: I don't see tests that show what happens when:
1. a client tries to access with http
2. a client tries to access with https his certificate is not from the same CA, etc... 
that is, negative tests

Comment (gennadylaventman) [reply pkg/server/server_test.go:683]: Will add such tests

**Comment by tock-ibm** [comment]
`scripts/tlsGen.sh` Is an empty file

File: pkg/server/testutils/crypto_utils.go [hunk 11]
```diff
@@ -215,21 +215,13 @@ func GenerateTestClientCrypto(t *testing.T, names []string, withIntermediateCA .
...
+		pemCert, privKey, err := IssueCertificate("BCDB cert for "+name, "127.0.0.1", keyPair)
```
Comment (tock-ibm) [inline pkg/server/testutils/crypto_utils.go:218]: BCDB -> Orion

File: pkg/server/server_test.go [hunk 12]
```diff
@@ -640,11 +725,233 @@ func TestServerWithFailureScenarios(t *testing.T) {
...
+	require.NoError(t, err)
+
+	err = verifier.Verify(userBytes, user.GetSignature())
+	require.NoError(t, err)
+
+	require.True(t, user.GetResponse().GetUser() != nil &&
+		user.GetResponse().GetUser().GetId() == "testUser")
+}
+
+func TestSyncTxWithServerHttsClientHttp(t *testing.T) {
```
Comment (tock-ibm) [inline pkg/server/server_test.go:786]: TestSyncTxWithServerHttpsClientHttp

File: pkg/server/server_test.go [hunk 13]
```diff
@@ -640,11 +725,233 @@ func TestServerWithFailureScenarios(t *testing.T) {
...
+	httpResp, err := env.client.SubmitTransaction(constants.PostUserTx,
+		&types.UserAdministrationTxEnvelope{
+			Payload:   userTx,
+			Signature: testutils.SignatureFromTx(t, env.adminSigner, userTx),
+		},
+		30*time.Second, //Sync
+	)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusBadRequest, httpResp.StatusCode)
+	httpResp.Body.Close()
```
Comment (tock-ibm) [inline pkg/server/server_test.go:827]: check also the content of the body?

File: config/config_test.go [hunk 14]
```diff
@@ -30,6 +30,18 @@ var expectedLocalConfig = &LocalConfiguration{
 			Block:                     100,
 		},
 		LogLevel: "info",
+		TLS: TLSConf{
+			Enabled:               false,
+			ClientAuthRequired:    false,
+			ServerCertificatePath: "./testdata/tls/server/server.cert",
+			ServerKeyPath:         "./testdata/tls/server/server.key",
+			ClientCertificatePath: "./testdata/tls/client/client.cert",
+			ClientKeyPath:         "./testdata/tls/client/client.key",
```
[the remark this answers is not in the document]
Comment (gennadylaventman) [reply config/config_test.go:39]: Just copied from `Replication.TLS` config. In this case need to change in both places...

File: config/testdata/config.yml [hunk 15]
```diff
@@ -41,6 +41,28 @@ server:
...
+    # Require client certificates / mutual TLS for inbound connections.
+    clientAuthRequired: false
+    # X.509 certificate used for TLS server
+    serverCertificatePath: ./testdata/tls/server/server.cert
+    # Private key for TLS server
+    serverKeyPath: ./testdata/tls/server/server.key
+    # X.509 certificate used for creating TLS client connections.
+    clientCertificatePath: ./testdata/tls/client/client.cert
+    # Private key used for creating TLS client connections.
+    clientKeyPath: ./testdata/tls/client/client.key
```
[the remark this answers is not in the document]
Comment (gennadylaventman) [reply config/testdata/config.yml:55]: Ok.

File: config/config_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (gennadylaventman) [reply config/config_test.go:39]: Removed

**Review by tock-ibm — APPROVED** [review]

**Pull request merged**